src/compiler/scanner.ts

```diff
@@ -2801,7 +2811,10 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
                         scanGroupName(/*isReference*/ true);
                         scanExpectedChar(CharacterCodes.greaterThan);
                     }
-                    else if (unicodeMode) {
+                    else {
+                        // This is actually allowed in Annex B if there are no named capturing groups in the regex,

```

Comment:
Could we keep track of whether we encountered a `(?<` during `reScanSlashToken` and add an entry to the `RegularExpressionFlags` enum? The spec passes `NamedCaptureGroups` as a production parameter just as it does for `UnicodeMode` and `UnicodeSetsMode`, but only ever passes it as `~NamedCaptureGroups` in [Annex B](https://tc39.es/ecma262/#sec-parsepattern-annexb).

Reply:
Oh, didn’t think of this clever but dirty solution.
I can implement this but… the point is still about how much Annex B things we are going to respect. If we allow `\k` then should we allow `\u` and `\x` (and also `\8` and `\9` inside character classes) too? (https://github.com/microsoft/TypeScript/pull/58320#discussion_r1581764104)

Reply:
I think it's better to error on `\u`, `\x`, `\8`, `\9` because in those cases you are more likely to have actually meant something different. Writing `\k` when there are no named capture groups is far less ambiguous.

Reply:
If then I would personally lean towards linting also `\1`~`\7` inside character classes to prevent [this kind of mistakes](https://github.com/microsoft/TypeScript/issues/58275#issuecomment-2068174007). Or perhaps the opposite, only outside character classes. I don’t know. I understand that you don’t want the syntax checking to be too breaky, but someone must find this useful.

Reply:
Every decimal escape in a character class I've ever seen has been a bug, so it makes sense to error for that case.

Reply:
Ah, but this would revert https://github.com/microsoft/TypeScript/pull/58295/files#diff-e19070e6b0ecc4de5f7d4bc6b42a9a71c8f23ce8ce155f7bd7cd48763170e506R1528-R1530. See if this is fine for you.